### ISSUE TITLE
forward_result::merge(): operation cannot be completed due to invalid argument sizes in test_resharding_by_murmur3_decrease

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#10684]
Seen in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-debug/15/artifact/logs-heavy.debug.000/1653777938143_resharding_test.py%3A%3ATestReshardingVariants%3A%3Atest_resharding_by_murmur3_decrease%5B4-TimeWindowCompactionStrategy-15%5D/node1.log
```
Scylla version 5.1.dev-0.20220527.e363febeb15e with build-id e7cfdadd5f32e4c92e5d19b7c5ffa712ed9effc5 starting ...
...
INFO  2022-05-28 22:43:10,668 [shard 0] init - Scylla version 5.1.dev-0.20220527.e363febeb15e initialization completed.
ERROR 2022-05-28 22:44:15,329 [shard 1] query - forward_result::merge(): operation cannot be completed due to invalid argument sizes. this.query_results.size(): 1 other.query_results.size(): 0 reduction_types.size(): 1, at: 0x13bcf010 0x1bccc12d 0x1bccbded 0x1bcce3e3 0x1bccf272 0x1adc00ab 0x179de2d8 0x16a658d1 0x16a6555c 0x16a652a3 0x16a65138 0x16a6502d 0x16a815d0 0x16a8140b 0x16a81202 0x16a8108a 0x16a80e36 0x1aeb5c6e 0x1aebc2e6 0x1aec0f52 0x1af99796 0x1af983f2 0x1af982a2 0x1af978f2 0x1452924a 0x1adc408b /jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.ccm/scylla-repository/e363febeb15e11bf5230cbb29714e22204cec269/scylla/libreloc/libpthread.so.0+0x92a4 /jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.ccm/scylla-repository/e363febeb15e11bf5230cbb29714e22204cec269/scylla/libreloc/libc.so.6+0x100322
   --------
   seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<query::forward_result>&&)>, seastar::future<query::forward_result>::then_wrapped_nrvo<seastar::future<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<query::forward_result>&&)> >(seastar::noncopyable_function<seastar::future<void> (seastar::future<query::forward_result>&&)>&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::noncopyable_function<seastar::future<void> (seastar::future<query::forward_result>&&)>&, seastar::future_state<query::forward_result>&&)#1}, query::forward_result>
Aborting on shard 1.
Backtrace:
  0x13bcf010
  0x1b02630d
  0x1b025f0c
  0x1ae96f74
  0x1aec2ace
  0x1af9743c
  0x1af9768d
  0x1af974ba
  0x7fe609e7ea1f
  /jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.ccm/scylla-repository/e363febeb15e11bf5230cbb29714e22204cec269/scylla/libreloc/libc.so.6+0x3d2a1
  /jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.ccm/scylla-repository/e363febeb15e11bf5230cbb29714e22204cec269/scylla/libreloc/libc.so.6+0x268a3
  0x1adc018b
  0x179de2d8
  0x16a658d1
  0x16a6555c
  0x16a652a3
  0x16a65138
  0x16a6502d
  0x16a815d0
  0x16a8140b
  0x16a81202
  0x16a8108a
  0x16a80e36
  0x1aeb5c6e
  0x1aebc2e6
  0x1aec0f52
  0x1af99796
  0x1af983f2
  0x1af982a2
  0x1af978f2
  0x1452924a
  0x1adc408b
  /jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.ccm/scylla-repository/e363febeb15e11bf5230cbb29714e22204cec269/scylla/libreloc/libpthread.so.0+0x92a4
  /jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.ccm/scylla-repository/e363febeb15e11bf5230cbb29714e22204cec269/scylla/libreloc/libc.so.6+0x100322
```

Decoded:
```
[Backtrace #0]
__interceptor_backtrace.part.0 at crtstuff.c:?
void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at ./build/debug/seastar/./seastar/include/seastar/util/backtrace.hh:59
seastar::backtrace_buffer::append_backtrace() at ./build/debug/seastar/./seastar/src/core/reactor.cc:750
seastar::print_with_backtrace(seastar::backtrace_buffer&, bool) at ./build/debug/seastar/./seastar/src/core/reactor.cc:780
seastar::print_with_backtrace(char const*, bool) at ./build/debug/seastar/./seastar/src/core/reactor.cc:792
seastar::sigabrt_action() at ./build/debug/seastar/./seastar/src/core/reactor.cc:3671
operator() at ./build/debug/seastar/./seastar/src/core/reactor.cc:3653
__invoke at ./build/debug/seastar/./seastar/src/core/reactor.cc:3649
?? ??:0
/data/scylla-s3-reloc.cache/by-build-id/e7cfdadd5f32e4c92e5d19b7c5ffa712ed9effc5/extracted/scylla/libexec/scylla/libreloc/libc.so.6 0x3d2a1 
/data/scylla-s3-reloc.cache/by-build-id/e7cfdadd5f32e4c92e5d19b7c5ffa712ed9effc5/extracted/scylla/libexec/scylla/libreloc/libc.so.6 0x268a3 
seastar::on_internal_error(seastar::logger&, std::basic_string_view<char, std::char_traits<char> >) at ./build/debug/seastar/./seastar/src/core/on_internal_error.cc:39
query::forward_result::merge(query::forward_result const&, std::vector<query::forward_request::reduction_type, std::allocator<query::forward_request::reduction_type> > const&) at ./query.cc:410
operator() at ./service/forward_service.cc:211
operator()<seastar::future<query::forward_result> > at ././seastar/include/seastar/core/map_reduce.hh:171
seastar::future<void> seastar::futurize<seastar::future<void> >::invoke<seastar::map_reduce<boost::range_detail::integer_iterator<unsigned int>, seastar::sharded<service::forward_service>::map_reduce0<service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_3, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4>(service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_3, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4)::{lambda(unsigned int)#1}, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4>(boost::range_detail::integer_iterator<unsigned int>, boost::range_detail::integer_iterator<unsigned int>, seastar::sharded<service::forward_service>::map_reduce0<service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_3, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4>(service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_3, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4)::{lambda(unsigned int)#1}&&, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4)::{lambda(auto:1)#1}&, seastar::future<query::forward_result> >(seastar::map_reduce<boost::range_detail::integer_iterator<unsigned int>, seastar::sharded<service::forward_service>::map_reduce0<service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_3, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4>(service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_3, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4)::{lambda(unsigned int)#1}, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4>(boost::range_detail::integer_iterator<unsigned int>, boost::range_detail::integer_iterator<unsigned int>, seastar::sharded<service::forward_service>::map_reduce0<service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_3, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4>(service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_3, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4)::{lambda(unsigned int)#1}&&, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4)::{lambda(auto:1)#1}&, seastar::future<query::forward_result>&&) at ././seastar/include/seastar/core/future.hh:2141
operator() at ././seastar/include/seastar/core/future.hh:1658
seastar::noncopyable_function<seastar::future<void> (seastar::future<query::forward_result>&&)>::direct_vtable_for<seastar::future<query::forward_result>::then_wrapped_maybe_erase<true, seastar::future<void>, seastar::map_reduce<boost::range_detail::integer_iterator<unsigned int>, seastar::sharded<service::forward_service>::map_reduce0<service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_3, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4>(service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_3, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4)::{lambda(unsigned int)#1}, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4>(boost::range_detail::integer_iterator<unsigned int>, boost::range_detail::integer_iterator<unsigned int>, seastar::sharded<service::forward_service>::map_reduce0<service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_3, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4>(service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_3, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4)::{lambda(unsigned int)#1}&&, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4)::{lambda(auto:1)#1}>(seastar::map_reduce<boost::range_detail::integer_iterator<unsigned int>, seastar::sharded<service::forward_service>::map_reduce0<service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_3, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4>(service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_3, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4)::{lambda(unsigned int)#1}, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4>(boost::range_detail::integer_iterator<unsigned int>, boost::range_detail::integer_iterator<unsigned int>, seastar::sharded<service::forward_service>::map_reduce0<service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_3, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4>(service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_3, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4)::{lambda(unsigned int)#1}&&, std::optional<query::forward_result>, service::forward_service::dispatch_to_shards(query::forward_request, std::optional<tracing::trace_info>)::$_4)::{lambda(auto:1)#1}&&)::{lambda(seastar::future<query::forward_result>&&)#1}>::call(seastar::noncopyable_function<seastar::future<void> (seastar::future<query::forward_result>&&)> const*, seastar::future<query::forward_result>&&) at ././seastar/include/seastar/util/noncopyable_function.hh:124
seastar::noncopyable_function<seastar::future<void> (seastar::future<query::forward_result>&&)>::operator()(seastar::future<query::forward_result>&&) const at ././seastar/include/seastar/util/noncopyable_function.hh:209
operator() at ././seastar/include/seastar/core/future.hh:1674
void seastar::futurize<seastar::future<void> >::satisfy_with_result_of<seastar::future<query::forward_result>::then_wrapped_nrvo<seastar::future<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<query::forward_result>&&)> >(seastar::noncopyable_function<seastar::future<void> (seastar::future<query::forward_result>&&)>&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::noncopyable_function<seastar::future<void> (seastar::future<query::forward_result>&&)>&, seastar::future_state<query::forward_result>&&)#1}::operator()(seastar::internal::promise_base_with_type<void>&&, seastar::noncopyable_function<seastar::future<void> (seastar::future<query::forward_result>&&)>&, seastar::future_state<query::forward_result>&&) const::{lambda()#1}>(seastar::internal::promise_base_with_type<void>&&, seastar::future<query::forward_result>::then_wrapped_nrvo<seastar::future<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<query::forward_result>&&)> >(seastar::noncopyable_function<seastar::future<void> (seastar::future<query::forward_result>&&)>&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::noncopyable_function<seastar::future<void> (seastar::future<query::forward_result>&&)>&, seastar::future_state<query::forward_result>&&)#1}::operator()(seastar::internal::promise_base_with_type<void>&&, seastar::noncopyable_function<seastar::future<void> (seastar::future<query::forward_result>&&)>&, seastar::future_state<query::forward_result>&&) const::{lambda()#1}&&) at ././seastar/include/seastar/core/future.hh:2126
operator() at ././seastar/include/seastar/core/future.hh:1673
seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<query::forward_result>&&)>, seastar::future<query::forward_result>::then_wrapped_nrvo<seastar::future<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<query::forward_result>&&)> >(seastar::noncopyable_function<seastar::future<void> (seastar::future<query::forward_result>&&)>&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::noncopyable_function<seastar::future<void> (seastar::future<query::forward_result>&&)>&, seastar::future_state<query::forward_result>&&)#1}, query::forward_result>::run_and_dispose() at ././seastar/include/seastar/core/future.hh:773
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at ./build/debug/seastar/./seastar/src/core/reactor.cc:2344
seastar::reactor::run_some_tasks() at ./build/debug/seastar/./seastar/src/core/reactor.cc:2757
seastar::reactor::do_run() at ./build/debug/seastar/./seastar/src/core/reactor.cc:2926
operator() at ./build/debug/seastar/./seastar/src/core/reactor.cc:4138
void std::__invoke_impl<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_89&>(std::__invoke_other, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_89&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/invoke.h:61
std::enable_if<is_invocable_r_v<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_89&>, void>::type std::__invoke_r<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_89&>(seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_89&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/invoke.h:111
std::_Function_handler<void (), seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_89>::_M_invoke(std::_Any_data const&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/std_function.h:290
std::function<void ()>::operator()() const at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/std_function.h:590
seastar::posix_thread::start_routine(void*) at ./build/debug/seastar/./seastar/src/core/posix.cc:60
```

Core dump here: https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-debug/15/artifact/logs-heavy.debug.000/1653777938143_resharding_test.py%3A%3ATestReshardingVariants%3A%3Atest_resharding_by_murmur3_decrease%5B4-TimeWindowCompactionStrategy-15%5D/node1-reactor-1.20719.1653777855.core.gz

Relocatable package here: http://downloads.scylladb.com/unstable/scylla/master/relocatable/2022-05-27T03:59:52Z/scylla-debug-x86_64-package.tar.gz

**Comment by bhalevy** [comment]
Seen also in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-debug/15/artifact/logs-full.debug.031/1653775490324_materialized_views_test.py%3A%3ATestInterruptBuildProcess%3A%3Atest_interrupt_build_process_with_resharding_half_to_max_test/node1.log
apparently on all nodes

**Comment by psarna** [comment]
@havaker please take a look, it's the count(*) code

**Comment by psarna** [comment]
I'll try to reproduce this fail on master, and if I somehow succeed, I'll run again with https://github.com/scylladb/scylla/pull/10602 .

**Comment by psarna** [comment]
My rationale is that https://github.com/scylladb/scylla/pull/10602 is a suspected miscompile, because the code which produces forward request results explicitly checks if the returned results are not empty, and the assertion from this issue claims that the results were empty anyway. So it's either a miscompilation or data being incorrectly passed via rpc, which is unlikely, because it's all simply generated with our idl.

**Comment by psarna** [comment]
Unfortunately, when I run this dtest locally (or any other similar dtest) in debug mode, it always times out with `FAILED resharding_test.py::TestReshardingVariants::test_resharding_by_murmur3_increase[1-TimeWindowCompactionStrategy-15] - cassandra.OperationTimedOut: errors={'Connection defunct by heartbeat': 'Client request timeout. See Session.execute[_async](timeout)'}, last_host=127.0.77.1:9042
`

**Comment by psarna** [comment]
Ok, I managed to run another test from this family (murmur3_decrease), and it succeeded with #10602. Now I'll check if it fails on master.

**Comment by psarna** [comment]
Nope, the `murmur3_decrease` keeps succeeding both on master and with the decoroutinization patch.

**Comment by bhalevy** [comment]
Other than `forward_service::dispatch_to_shards` being quite inefficient due to copying of `req` into the mapper lambda, and it being copied for each shard, I think the reducer better capture `req.reduction_types` rather than getting it view `&req`.

**Comment by psarna** [comment]
@bhalevy that's also done as part of the decoroutinization process in #10602. @avikivity would prefer not to merge the decoroutinization without solid proof that it fixes the problem and I sympathize with this approach (otherwise we can pessimize the code without good reason for it), but the problem is unfortunately not trivial to trigger.

**Comment by bhalevy** [comment]
Hit again in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-debug/21/testReport/materialized_views_test/TestInterruptBuildProcess/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split030___test_interrupt_build_process_with_resharding_half_to_max_test_2/

I suggest we fix this issue independently of https://github.com/scylladb/scylla/pull/10602

**Comment by psarna** [comment]
@bhalevy but #10602 is literally the fix :) Now we know for sure that the code with references was indeed illegal, and in my opinion uncoroutinizing the lambda is a far more human-readable solution than instantiating the references on the coroutine frame. This code will soon be overhauled by #10295 , but there's no harm (except some minimal rebasing effort) in merging #10602 before that - and #10602 also gets rid of a std::optional, which is a nice refactor.

**Comment by mykaul** [comment]
@eliransin - can you assign someone to follow up on Sarna's comment above and see that we make progress on this?

**Comment by DoronArazii** [comment]
@bhalevy do we still need to track it?

**Comment by bhalevy** [comment]
Apparently fixed by #11064
